sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldBeInitializedCorrectlyBase.cs

```diff
@@ -0,0 +1,210 @@
+                    }
+                },
+                SyntaxKind.ObjectCreationExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var assignment = (AssignmentExpressionSyntax)c.Node;
+
+                    // Ignore assignments within object initializers, they are
+                    // reported in the ObjectCreationExpression handler
+                    if (assignment.FirstAncestorOrSelf<InitializerExpressionSyntax>() == null &&
+                        IsTrackedPropertyName(assignment.Left) &&
+                        IsPropertyOnTrackedType(assignment.Left, c.SemanticModel) &&
+                        !IsAllowedValue(assignment.Right, c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, assignment.GetLocation()));
+                    }
+                },
+                SyntaxKind.SimpleAssignmentExpression);
+        }
+
+        /// <summary>
+        /// Tests if the provided expression is a property of the <see cref="TrackedType"/>. Override this method
+        /// when the <see cref="TrackedPropertyName"/> is a indexer for example.
+        /// </summary>
+        /// <returns>True when the parameter is a property of the <see cref="TrackedType"/>, otherwise false.</returns>
+        protected virtual bool IsPropertyOnTrackedType(ExpressionSyntax expression, SemanticModel semanticModel) =>
+            expression is MemberAccessExpressionSyntax memberAccess &&
+            memberAccess.Expression != null &&
+            IsTrackedType(memberAccess.Expression, semanticModel);
+
+        /// <summary>
+        /// Tests it the provided <paramref name="argumentList"/> contains initialization of the <see cref="TrackedPropertyName"/>
+        /// with an allowed value. Override this method when a constructor can initialize the property value with a allowed value.
+        /// </summary>
+        /// <returns>True when the <paramref name="argumentList"/> contains initialization of <see cref="TrackedPropertyName"/>
+        /// with allowed value or when this constructor sets <see cref="TrackedPropertyName"/> with the
+        /// allowed value by default, otherwise false.</returns>
+        protected virtual bool CtorInitializesTrackedPropertyWithAllowedValue(ArgumentListSyntax argumentList, SemanticModel semanticModel) =>
+            false;
+
+        /// <summary>
+        /// Tests if the <paramref name="expression"/> is <see cref="TrackedPropertyName"/>.
+        /// </summary>
+        /// <returns>True if the <paramref name="expression"/> is <see cref="TrackedPropertyName"/>,
+        /// otherwise false.</returns>
+        protected virtual bool IsTrackedPropertyName(ExpressionSyntax expression)
+        {
+            var memberAccess = expression as MemberAccessExpressionSyntax;
```

Comment:
[minor] As you are doing `(expression as IdentifierNameSyntax)` on the next line you could probably merge this line with the next one too.